File: BackUpForABook/UIControls/TRClasses/TRCenterLabel.m
//
//  TRCenterLabel.m
//  ATreeRingMap
//
//  Created by Hui Wang on 7/27/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "TRCenterLabel.h"
#import "Define.h"

// label group used to show messages at the tree ring map center
@implementation TRCenterLabel

-(id)init
{
    self = [super init];
    
    _metricInfo = [[ANText alloc] initWithFont:@"Helvetica" size:12 shadow:FALSE underline:FALSE];
    _metricValue = [[ANText alloc] initWithFont:@"Helvetica" size:20 shadow:TRUE underline:FALSE];
    _title = [[ANText alloc] initWithFont:@"Helvetica-Bold" size:12 shadow:FALSE underline:FALSE];
    
    _bkgColorHi = [UIColor colorWithRed:1.0 green:1.0 blue:1.0 alpha:.2];
    _bkgColorLo = [UIColor colorWithRed:1.0 green:.0 blue:.0 alpha:.1];
    
    return self;
}

// test method
-(void)fillTestData
{
    _metricInfo.text = [NSMutableString stringWithString:@"Your Heart Years: "];
    _metricInfo.textFillColor = [UIColor blackColor];
    _metricInfo.textStrokeColor = [UIColor darkGrayColor];
    
    _metricValue.text = [NSMutableString stringWithString:@"-20"];
    _metricValue.textFillColor = [UIColor redColor];
    _metricValue.textStrokeColor = [UIColor blackColor];
    
    _title.text = [NSMutableString stringWithString:@"Tree Of Life"];
    _title.textFillColor = [UIColor darkGrayColor];
    _title.textStrokeColor = [UIColor grayColor];
}

// paint
-(void)paint:(CGContextRef)ctx
      radius:(float)radius
      origin:(CGPoint)origin
{
    // (1) background
    [self paintGradient2:ctx radius:radius origin:origin];
    
    // (2) title
    CGSize titleSize = [_title.attributedText size];
    CGRect rect = CGRectMake(origin.x - .5 * titleSize.width, (origin.y + titleSize.height), titleSize.width, titleSize.height + 2);
    
    [_title paint:ctx rect:rect attributedText:_title.attributedText];
    
    // (3) metricValue
    CGSize metricValueSize = [_metricValue.attributedText size];
    
    rect = CGRectMake(origin.x - .5 * metricValueSize.width, (origin.y - .5 * metricValueSize.height), metricValueSize.width, metricValueSize.height + 2);
    
    [_metricValue paint:ctx rect:rect attributedText:_metricValue.attributedText];
    
    // (4) metric
    CGSize metricSize = [_metricInfo.attributedText size];
    
    rect = CGRectMake(origin.x - .5 * metricSize.width, (origin.y - metricValueSize.height - .5 * metricSize.height), metricSize.width, metricSize.height + 2);
    
    [_metricInfo paint:ctx rect:rect attributedText:_metricInfo.attributedText];
}


-(void)paintGradient2:(CGContextRef)ctx
              radius:(float)radius
              origin:(CGPoint)origin
{
    // gradient color
    size_t num_locations = 2;
    CGFloat locations[] = {.0, 1.0};
    
    // gradient color
    CGFloat gradientColors[8] = {.0, .0, 1.0, .7, 1.0, 1.0, 1.0, .8};
    
    // edge
    CGColorRef colorref = [_bkgColorLo CGColor];
    int numComponents = CGColorGetNumberOfComponents(colorref);
    if (numComponents == 4) {
        const CGFloat *components = CGColorGetComponents(colorref);
        CGFloat red     = components[0];
        CGFloat green = components[1];
        CGFloat blue   = components[2];
        CGFloat alpha = components[3];
   
        gradientColors[4] = red;
        gradientColors[5] = green;
        gradientColors[6] = blue;
        gradientColors[7] = alpha;
     }
    // center
    colorref = [_bkgColorHi CGColor];
    numComponents = CGColorGetNumberOfComponents(colorref);
    if (numComponents == 4) {
        const CGFloat *components = CGColorGetComponents(colorref);
        CGFloat red     = components[0];
        CGFloat green = components[1];
        CGFloat blue   = components[2];
        CGFloat alpha = components[3];
        
        gradientColors[0] = red;
        gradientColors[1] = green;
        gradientColors[2] = blue;
        gradientColors[3] = alpha;
    }
    
    CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
    CGGradientRef gradient = CGGradientCreateWithColorComponents(space, gradientColors, locations, num_locations);
    
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(0.0) endAngle:-DEGREES_TO_RADIANS(180.0) clockwise:FALSE];
    [aPath closePath];
    
    CGContextSaveGState(ctx);
    CGContextBeginPath(ctx);
    
    CGContextAddPath(ctx, aPath.CGPath);
    
    CGContextClip(ctx);
    
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 3.0), 3.0, [UIColor grayColor].CGColor);
    
    CGContextDrawRadialGradient(ctx, gradient, origin, 0.0, origin, radius, kCGGradientDrawsBeforeStartLocation);
    
    CGContextRestoreGState(ctx);
    CGColorSpaceRelease(space);
    CGGradientRelease(gradient);
}

-(void)paintGradient3:(CGContextRef)ctx
              radius:(float)radius
              origin:(CGPoint)origin
{
    // gradient color
    size_t num_locations = 3;
    CGFloat locations[] = {.0, .5, 1.0};
    
    // gradient color
    CGFloat gradientColors[12] = {.0, .0, 1.0, .7, 1.0, 1.0, 1.0, .8, .0, .0, 1.0, .7};
    
    // edge
    CGColorRef colorref = [_bkgColorHi CGColor];
    int numComponents = CGColorGetNumberOfComponents(colorref);
    if (numComponents == 4) {
        const CGFloat *components = CGColorGetComponents(colorref);
        CGFloat red     = components[0];
        CGFloat green = components[1];
        CGFloat blue   = components[2];
        CGFloat alpha = components[3];
        
        gradientColors[0] = red;
        gradientColors[1] = green;
        gradientColors[2] = blue;
        gradientColors[3] = alpha;
        
        gradientColors[8] = red;
        gradientColors[9] = green;
        gradientColors[10] = blue;
        gradientColors[11] = alpha;
    }
    // middle
    colorref = [_bkgColorLo CGColor];
    numComponents = CGColorGetNumberOfComponents(colorref);
    if (numComponents == 4) {
        const CGFloat *components = CGColorGetComponents(colorref);
        CGFloat red     = components[0];
        CGFloat green = components[1];
        CGFloat blue   = components[2];
        CGFloat alpha = components[3];
        
        gradientColors[4] = red;
        gradientColors[5] = green;
        gradientColors[6] = blue;
        gradientColors[7] = alpha;
    }
    
    CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
    CGGradientRef gradient = CGGradientCreateWithColorComponents(space, gradientColors, locations, num_locations);
    
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(0.0) endAngle:-DEGREES_TO_RADIANS(180.0) clockwise:FALSE];
    [aPath closePath];
    
    CGContextSaveGState(ctx);
    CGContextBeginPath(ctx);
    
    CGContextAddPath(ctx, aPath.CGPath);
    
    CGContextClip(ctx);
    
    CGContextDrawRadialGradient(ctx, gradient, origin, 0.0, origin, radius, kCGGradientDrawsBeforeStartLocation);
 
    CGContextRestoreGState(ctx);
    CGColorSpaceRelease(space);
    CGGradientRelease(gradient);
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/TRCenterLabel.swift
//
//  TRCenterLabel.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/24.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class TRCenterLabel: NSObject {
    var metricInfo: ANText?
    var metricValue: ANText?
    var title: ANText?
    var bkgColorHi: UIColor?
    var bkgColorLo: UIColor?
    
    
    
    override init() {
        super.init()
        metricInfo = ANText.init("Helvetica", size: 12, shadow: false, underline: false)
        metricValue = ANText.init("Helvetica", size: 20, shadow: true, underline: false)
        title = ANText.init("Helvetica-Bold", size: 12, shadow: false, underline: false)
        
        bkgColorHi = UIColor(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0)
        bkgColorLo = UIColor(red: 255/255.0, green: 207/255.0, blue: 207/255.0, alpha: 1.0)
    }
    
    // paint
    func paint(_ ctx: CGContext, radius: Float, origin: CGPoint,borderWidth: Float){
        paintBorder(ctx, radius: radius + borderWidth / 2.0, origin: origin,width: borderWidth)
        // 1. background
        paintGradient2(ctx, radius: radius, origin: origin)
        
        // 2. title
        let titleSize = title?.attributedText.size()
        var rect = CGRect(x: origin.x - 0.5 * titleSize!.width, y: origin.y + titleSize!.height, width: titleSize!.width, height: titleSize!.height + 2)
        title?.paint(ctx, rect: rect)
        
        // 3. metric value
        let metricValueSize = metricValue?.attributedText.size()
        rect = CGRect(x: origin.x - 0.5 * metricValueSize!.width, y: origin.y - 0.5 * metricValueSize!.height, width: metricValueSize!.width, height: metricValueSize!.height + 2)
        metricValue?.paint(ctx, rect: rect)
        
        // 4. metric
        let metricSize = metricInfo?.attributedText.size()
        rect = CGRect(x: origin.x - 0.5 * metricSize!.width, y: origin.y - metricValueSize!.height - 0.5 * metricSize!.height, width: metricSize!.width, height: metricSize!.height + 2)
        metricInfo?.paint(ctx, rect: rect)

    }
    
    // color border of center label
    fileprivate func paintBorder(_ ctx: CGContext,radius: Float, origin: CGPoint,width: Float){

        ctx.setStrokeColor(UIColor.white.cgColor)
        ctx.setLineWidth(CGFloat(width))
        ctx.addArc(center: origin, radius: CGFloat(radius), startAngle: CGFloat(DEGREETORADIANS(degree: 0.0)), endAngle:  (CGFloat(DEGREETORADIANS(degree: 180))), clockwise: true)
        ctx.drawPath(using: CGPathDrawingMode.stroke)
      
    }
    
    fileprivate func paintGradient2(_ ctx: CGContext,radius: Float, origin: CGPoint){
        // gradient color
        let num_locations = 2
        let locations: [CGFloat] = [0.0, 1.0]
        
        // gradient color
        var gradientColors: [CGFloat] = [0.0, 0.0, 1.0, 0.7, 1.0, 1.0, 1.0, 0.8]
        
        // edge
        var colorref = bkgColorLo?.cgColor
        var numComponents = colorref?.numberOfComponents
        if numComponents == 4{
            let components = colorref?.components
            let red = components?[0]
            let green = components?[1]
            let blue = components?[2]
            let alpha = components?[3]
            
            gradientColors[4] = red!
            gradientColors[5] = green!
            gradientColors[6] = blue!
            gradientColors[7] = alpha!
        }
        // center
        colorref = bkgColorHi?.cgColor
        numComponents = colorref?.numberOfComponents
        if numComponents == 4{
            let components = colorref?.components
            let red = components?[0]
            let green = components?[1]
            let blue = components?[2]
            let alpha = components?[3]
            
            gradientColors[0] = red!
            gradientColors[1] = green!
            gradientColors[2] = blue!
            gradientColors[3] = alpha!
        }
        
        let space = CGColorSpaceCreateDeviceRGB()
        let gradient = CGGradient.init(colorSpace: space, colorComponents: gradientColors, locations: locations, count: num_locations)
        
        let aPath = UIBezierPath()
        aPath.addArc(withCenter: origin, radius: CGFloat(radius), startAngle: -CGFloat(DEGREETORADIANS(degree: 0.0)), endAngle: -(CGFloat(DEGREETORADIANS(degree: 180))), clockwise: false)
        aPath.close()
        
        ctx.saveGState()
        ctx.beginPath()
        ctx.addPath(aPath.cgPath)
        ctx.clip()
        ctx.setShadow(offset: CGSize.init(width: 0, height: 3.0), blur: 3.0, color: UIColor.gray.cgColor)
        ctx.drawRadialGradient(gradient!, startCenter: origin, startRadius: 0.0, endCenter: origin, endRadius: CGFloat(radius), options: .drawsBeforeStartLocation)
        ctx.restoreGState()
    }

}
